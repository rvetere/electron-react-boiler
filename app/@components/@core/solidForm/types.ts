import { InputValue } from '.'
import { IValidationMessage } from './elements/infoSlot/validationDisplay/types'

export enum LabelPosition {
  LabelLeft = 'LabelLeft',
  LabelRight = 'LabelRight'
}

export interface IInputDefaultState {
  value?: InputValue
  isValid: boolean
  isTouched: boolean
  validationMessages: IValidationMessage[]
}
