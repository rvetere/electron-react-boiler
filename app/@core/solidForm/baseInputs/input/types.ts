import { InputValue, ISelectOption } from '@core/solidForm'
import { IValidationDisplayConfig, IValidationMessage } from '@core/solidForm/elements/infoSlot/validationDisplay/types'
import { IFormControllerInputProps } from '@core/solidForm/formController/types'
import { IInputDefaultState } from '@core/solidForm/types'
import { ValidatorFn } from '@core/solidForm/validators/types'
import React, { ReactNode } from 'react'

export interface IInputTextProps extends React.InputHTMLAttributes<HTMLInputElement> {
  name: string
  inputProps: IFormControllerInputProps
  validators?: ValidatorFn[]
  inverted?: boolean
  label?: ReactNode
  labelAppendix?: ReactNode
  hideValidation?: boolean
  infoSlot?: (validationConfig: IValidationDisplayConfig) => ReactNode
}

export type IInputTextState = IInputDefaultState

export interface IInputValidationPayload {
  value?: InputValue | ISelectOption[]
  validators?: ValidatorFn[]
  inputRef?: React.RefObject<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>
}

export interface IInputValidationResult {
  isValid: boolean
  messages: IValidationMessage[]
}
