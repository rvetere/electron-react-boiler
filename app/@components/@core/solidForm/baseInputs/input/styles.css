.wrapper {
  input {
    &:focus,
    &:active {
      border-color: $primaryColor;
    }
  }
}

.error {
  input {
    border-color: $errorColor;
  }
}
