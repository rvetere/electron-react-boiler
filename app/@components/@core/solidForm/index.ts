export { Checkbox } from './baseInputs/checkbox'
export { ICheckboxProps } from './baseInputs/checkbox/types'
export { InputText } from './baseInputs/input'
export { IInputTextProps } from './baseInputs/input/types'
export { TextArea } from './baseInputs/textArea'
export { ITextAreaProps } from './baseInputs/textArea/types'
export { RadioGroup } from './baseInputs/radioGroup'
export { IRadioGroupProps, IRadioProps } from './baseInputs/radioGroup/types'
export { Select } from './baseInputs/select'
export { getSelectedOption } from './baseInputs/select/helpers'
export { ISelectProps, ISelectOption } from './baseInputs/select/types'
export { IGroupType } from './baseInputs/select/selectPlugin/types'
export { RadioOption } from './baseInputs/radioGroup/radio'
export { IRadioOptionProps } from './baseInputs/radioGroup/radio/types'
export { InputPassword } from './extendedInputs/inputPassword'
export { RadioRow } from './extendedInputs/radioRow'
export { InputWithLimit, FieldWithLimitType } from './extendedInputs/inputWithLimit'
export { IInputTextWithLimitProps, ITextAreaWithLimitProps } from './extendedInputs/inputWithLimit/types'
export { FormController } from './formController'
export {
  IFormControllerProps,
  IFormControllerActions,
  IFormValues,
  FormSubmitFn,
  FormResetFn,
  IFormControllerInputProps,
  IFormControllerFields,
  IFormControllerField
} from './formController/types'
export { LabelPosition } from './types'

export type InputValue = string[] | string | number | undefined
