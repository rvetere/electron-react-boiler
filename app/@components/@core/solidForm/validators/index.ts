import isPostalCode from './helpers/isPostalCode'
import { IValidatorResult } from './types'

export const RequiredMsgDefault = 'Required'

export const longerThan5 = (value: string): IValidatorResult => ({
  isValid: value.length > 5,
  message: { message: 'More than 5 characters' }
})

export const containsUpperCase = (value: string): IValidatorResult => ({
  isValid: /[A-Z]/.test(value),
  message: { message: 'Min 1 UpperCase' }
})

export const isZip = (value: string): IValidatorResult => ({
  isValid: isPostalCode(value),
  message: { message: 'Invalid zip' }
})
