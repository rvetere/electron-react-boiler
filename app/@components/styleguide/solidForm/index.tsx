import gridCss from '@postcss/utils/grid.css'
import React from 'react'
import Form from './sample'

export default (): JSX.Element => (
  <>
    <h2>Solid Form</h2>
    <div className={gridCss.col12}>
      <Form />
    </div>
  </>
)
