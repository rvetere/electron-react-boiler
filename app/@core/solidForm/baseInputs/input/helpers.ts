import { InputValue, ISelectOption } from '@core/solidForm'
import { ValidatorFn } from '@core/solidForm/validators/types'
import { ICheckboxProps } from '../checkbox/types'
import { IRadioGroupProps } from '../radioGroup/types'
import { ISelectProps } from '../select/types'
import { ITextAreaProps } from '../textArea/types'
import { IInputTextProps, IInputValidationPayload, IInputValidationResult } from './types'

export const applyValidators = (value: InputValue | ISelectOption[], validators: ValidatorFn[]): IInputValidationResult =>
  validators.reduce(
    (result: IInputValidationResult, validator: ValidatorFn): IInputValidationResult => {
      const { isValid, message } = validator(value)

      if (result.isValid !== false) {
        result.isValid = isValid
      }

      if (message && !isValid) {
        result.messages = [...result.messages, message]
      }

      return result
    },
    { isValid: true, messages: [] }
  )

export const getNativeValidation = ({ current }: React.RefObject<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>): IInputValidationResult => {
  return {
    isValid: current ? current.validity.valid : true,
    messages:
      current && !current.validity.valid
        ? [
            {
              message: current.validationMessage
            }
          ]
        : []
  }
}

export const validateInput = ({ value, validators, inputRef }: IInputValidationPayload): IInputValidationResult => {
  const { messages, isValid } = validators ? applyValidators(value, validators) : { isValid: true, messages: [] }
  const { messages: nativeMessages, isValid: nativeIsValid } = inputRef ? getNativeValidation(inputRef) : { isValid: true, messages: [] }

  return {
    isValid: isValid && nativeIsValid,
    messages: [...messages, ...nativeMessages]
  }
}

export const hasValidation = (props: ICheckboxProps | IInputTextProps | IRadioGroupProps | ISelectProps | ITextAreaProps): boolean => {
  // @ts-ignore
  return props.required || (props.validators && props.validators.length > 0)
}
