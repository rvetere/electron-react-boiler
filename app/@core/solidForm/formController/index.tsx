import logger from '@helpers/logger'
import noop from '@helpers/noop'
import { PureComponent, ReactNode } from 'react'
import { getFormValues, isFormValid, updateFormFields } from './helpers'
import { IFormControllerField, IFormControllerPayload, IFormControllerProps, IFormControllerState } from './types'

export class FormController extends PureComponent<IFormControllerProps, IFormControllerState> {
  public readonly state: IFormControllerState = { fields: {}, isLoading: false }

  public render(): ReactNode {
    const { children, className, hideWrappingFormTag = false } = this.props
    const childrenJsx = children({
      inputProps: {
        updateForm: this.updateForm,
        registerInput: this.registerInput,
        deRegisterInput: this.deRegisterInput
      },
      submitForm: this.submitForm,
      resetForm: this.resetForm,
      formState: this.state
    })

    if (hideWrappingFormTag) {
      return childrenJsx
    }

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        {childrenJsx}
      </form>
    )
  }

  private updateForm = (submittedData: IFormControllerPayload): void => {
    const { onUpdate = noop } = this.props
    const { fields } = this.state

    if (fields.hasOwnProperty(submittedData.name)) {
      if (fields[submittedData.name].value !== submittedData.value) {
        this.setState(updateFormFields(submittedData), () => onUpdate(getFormValues(this.state.fields)))
      }
    } else {
      logger.warn(`Unregistered form input ${submittedData.name} attempted to update.`)
    }
  }

  private registerInput = (submittedData: IFormControllerPayload): void => {
    this.setState(updateFormFields(submittedData))
  }

  private deRegisterInput = (submittedData: IFormControllerPayload): void => {
    const { fields } = this.state
    delete fields[submittedData.name]
    this.setState({ fields })
  }

  private handleSubmit = (e: React.FormEvent): void => {
    e.preventDefault()
    this.submitForm()
  }

  private submitForm = async (): Promise<void> => {
    const { onSubmit = noop } = this.props
    const { fields } = this.state

    this.touchForm()
    if (isFormValid(fields)) {
      this.setState({ isLoading: true })
      try {
        await onSubmit(getFormValues(fields), this.resetForm)
        this.setState({ isLoading: false })
      } catch (e) {
        logger.error('Could not submit form!', e)
        this.setState({ isLoading: false })
      }
    } else {
      this.handleInvalid()
    }
  }

  private handleInvalid = (): void => {
    const { onInvalid = noop } = this.props
    onInvalid()
  }

  private touchForm = (): void => {
    const { fields } = this.state

    Object.values(fields).forEach(({ touch }: IFormControllerField): void => touch())
  }

  private resetForm = (): void => {
    const { onReset = noop } = this.props
    const { fields } = this.state

    Object.values(fields).forEach(({ reset }: IFormControllerField): void => reset())
    onReset()
  }
}
