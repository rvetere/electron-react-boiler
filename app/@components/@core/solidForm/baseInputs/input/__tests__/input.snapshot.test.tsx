/* eslint-env jest */
import { containsUpperCase } from '@core/solidForm/validators'
import { getByClass } from '@testing/classQueries'
import { cleanup, fireEvent, render, waitForElement } from 'react-testing-library'
import { InputText } from '..'
import { getInputMock } from './__mocks__/input.mock'

describe('InputText', () => {
  afterEach(cleanup)
  const mockProps = getInputMock()

  test('InputText component should render correctly', () => {
    const { container } = render(<InputText {...mockProps} />)

    expect(container).toMatchSnapshot()
    expect(mockProps.inputProps.registerInput).toBeCalledTimes(1)
  })

  test('InputText should validate', async () => {
    const validators = [containsUpperCase]

    const { container } = render(<InputText {...mockProps} validators={validators} />)

    const elem = container.querySelector('input')
    fireEvent.blur(elem!)
    fireEvent.change(elem!, { target: { value: 'must fail' } })

    const validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))

    expect(validationDisplay).toBeTruthy()
    expect(mockProps.inputProps.updateForm).toBeCalledTimes(1)
  })
})
