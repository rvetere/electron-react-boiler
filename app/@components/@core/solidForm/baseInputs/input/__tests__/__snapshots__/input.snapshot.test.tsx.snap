// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText InputText component should render correctly 1`] = `
<div>
  <div
    class="formField wrapper optional"
  >
    <label
      class="formLabel label"
    >
      <span>
        Input
      </span>
      <input
        aria-describedby="inputValidationMessage"
        name="input"
        value=""
      />
    </label>
  </div>
</div>
`;
